boskos/server/server.go

```diff
@@ -0,0 +1,303 @@
+// In: name - name of the target resource
+//     dest - destination state of the resource
+//     owner - owner of the resource
+// Out: error if owner does not match, or resource does not exist
+func (r *Ranch) Release(name string, dest string, owner string) error {
+	r.lock.Lock()
+	defer r.lock.Unlock()
+
+	for idx := range r.Resources {
+		res := &r.Resources[idx]
+		if name == res.Name {
+			if owner != res.Owner {
+				return &OwnerNotMatch{res.Owner, owner}
+			}
+			res.LastUpdate = time.Now()
+			res.Owner = ""
+			res.State = dest
+			return nil
+		}
+	}
+
+	return &ResourceNotFound{name}
+}
+
+// Update updates the timestamp of a target resource.
+// In: name - name of the target resource
+//     state - current state of the resource
+//     owner - current owner of the resource
+// Out: error if state or owner does not match, or resource does not exist
+func (r *Ranch) Update(name string, owner string, state string) error {
+	r.lock.Lock()
+	defer r.lock.Unlock()
+
+	for idx := range r.Resources {
+		res := &r.Resources[idx]
+		if name == res.Name {
+			if owner != res.Owner {
+				return &OwnerNotMatch{res.Owner, owner}
+			}
+
+			if state != res.State {
+				return &StateNotMatch{res.State, state}
+			}
+			res.LastUpdate = time.Now()
+			return nil
+		}
+	}
+
+	return &ResourceNotFound{name}
+}
+
+// Reset unstucks a type of stale resource to a new state.
+// In: rtype - type of the resource
+//     state - current state of the resource
+//     expire - duration before resource's last update
+//     dest - destination state of expired resources
+// Out: map of resource name - resource owner
+func (r *Ranch) Reset(rtype string, state string, expire time.Duration, dest string) map[string]string {
+	r.lock.Lock()
+	defer r.lock.Unlock()
+
+	ret := make(map[string]string)
+
+	for idx := range r.Resources {
+		res := &r.Resources[idx]
+		if rtype == res.Type && state == res.State && res.Owner != "" {
+			if time.Now().Sub(res.LastUpdate) > expire {
+				res.LastUpdate = time.Now()
+				ret[res.Name] = res.Owner
+				res.Owner = ""
+				res.State = dest
+			}
+		}
+	}
+
+	return ret
+}
+
+// LogStatus outputs current status of all resources
+func (r *Ranch) LogStatus() {
+	r.lock.RLock()
+	defer r.lock.RUnlock()
+
+	for _, res := range r.Resources {
+		resJSON, _ := json.Marshal(res)
+		logrus.Infof("Current Resources : %v", string(resJSON))
+	}
+}
+
+// SyncConfig updates resource list from a file
+func (r *Ranch) SyncConfig(config string) error {
+	r.lock.Lock()
+	defer r.lock.Unlock()
+
+	file, err := ioutil.ReadFile(config)
+	if err != nil {
+		return err
+	}
+
+	var data []common.Resource
+	err = json.Unmarshal(file, &data)
+	if err != nil {
+		return err
+	}
+
+	return r.syncConfigHelper(data)
+}
+
+func (r *Ranch) syncConfigHelper(data []common.Resource) error {
+	// delete non-exist resource
+	valid := 0
+	for _, res := range r.Resources {
+		// If currently busy, yield deletion to later cycles.
```

Comment:
This is very cool design -- how we add and remove items from the ranch. Can you better document these functions? Also is this in your design doc?